modules/world/world.cpp

```diff
@@ -41,6 +43,50 @@ World::World(const std::shared_ptr<World>& world) :
   std::signal(SIGSEGV, modules::commons::SegfaultHandler);
 }
 
+void World::Step(const float& delta_time) {
+  const float inc_world_time = world_time_ + delta_time;
+  PlanAgents(delta_time);
+  Execute(inc_world_time);
+  world_time_ = inc_world_time;
+}
+
+void World::PlanAgents(const float& delta_time) {
+  UpdateAgentRTree();
+  WorldPtr current_world(this->Clone());
+  const float inc_world_time = world_time_ + delta_time;
+  for (auto agent : agents_) {
+    ObservedWorld observed_world(current_world, agent.first);
+    agent.second->PlanBehavior(delta_time, observed_world);
+    if (agent.second->GetBehaviorStatus() == BehaviorStatus::VALID)
+      agent.second->PlanExecution(inc_world_time);
+  }
+}
+
+void World::Execute(const float& world_time) {
+  using models::dynamic::StateDefinition::TIME_POSITION;
+  for (auto agent : agents_) {
+    if (agent.second->GetBehaviorStatus() == BehaviorStatus::VALID &&
+        agent.second->GetExecutionStatus() == ExecutionStatus::VALID) {
+      agent.second->UpdateStateAction();
+      // make sure all agents have the same world time
+      // otherwise the simulation is not correct
+      const auto& agent_state = agent.second->GetCurrentState();
+      assert(fabs(agent_state(TIME_POSITION) - world_time) < 0.01);
+    }
+  }
+  RemoveInvalidAgents();
+}
+
+WorldPtr World::GetWorldAtTime(const float& world_time) const {
+  WorldPtr current_world_state(this->Clone());
+  for (auto agent : current_world_state->GetAgents()) {
+    if (agent.second->GetBehaviorStatus() == BehaviorStatus::VALID)
+      agent.second->PlanExecution(world_time);
+    agent.second->UpdateStateAction();
```

Comment:
if Execution Valid should be checked here

Reply:
I do not think so, as PlanExecution calls the exxecution model of the agent.